### PR TITLE
Make Directionality.of non-null

### PR DESCRIPTION
**Pull request by goderbauer**
## Description

In the majority of cases we assume in the framework that `Directionality.of` returns a non-null value. This PR changes the API of that method to guarantee a non-null return value (and throw if no Directionality can be found) to clean up those call sides. Additionally, it adds a new `Directionality.maybeOf` method for those handful of cases, we a null return value is actually ok.

I think overall this leads to cleaner code because in the cases where you don't expect `Directionality.of` to return null you can remove the `!`.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Haider2810** [comment]
Stop Sending Emails Twat

**Comment by goderbauer** [comment]
Customer tests for flutter_svg are failing on this change.

**Comment by goderbauer** [comment]
In google this change uncovered 6 buggy tests, which were accidentally passing with Directionality.of returning null even though they shouldn't. Fixes for that a purely in the test by adding a Directionality widget to them.

One widget in google3 needs to have `Directionality.of` replaced by `Directionality.maybeOf`.

**Comment by goderbauer** [comment]
@dnfield @Hixie @gspencergoog Curious to hear what you think about this change. I think overall it improves the readability of the code by removing a lot of `!` in places where we just assume `Directionality.of` to be non-null (and it also now throws a nice error message in case somebody forgot to include a Directionality widget). However, it is a breaking change. Running this through our customer tests, I found two instances that need to replace their usage of `Directionality.of` with the newly introduced `Directionality.maybeOf` API:

* In flutter_svg, there is one valid call of `Directionality.of` that is legitimately ok  to return null.
* In google there's also one similar call like this.

At the same time, this change also uncovered 6 buggy tests in google, where the test authors forgot to include a Directionality widget (even though they should have). 

Overall, I am in favor of this change, but like I said, it is slightly breaking.

**Comment by dnfield** [comment]
Some of our `.of` API has a `nullOk` flag, e.g. `Brightness.of`, `MediaQuery.of`. Would we want to apply similar refactors to those?

If not, I think we might want to just add a `nullOk` here to make it consistent.

If so, I think we should evaluate when it's ok for these to be null or not. It seems like we have API usages that expect this to be nullable. I'd probably be more in favor of adding some assert(s) to this so that tests have to opt in to a nullOk. I'm having trouble deciding whether this is really the right way to go - I can think of cases where you'd ask about the direcitionality of your context and be ok with it not existing, and other cases where you wouldn't. You can make that clear already by using `!` in your call. I'm a little concerned about adding yet another way to deal with missing inherited widgets in the tree (e.g. always allow null, allow null if `nullOk`, allow null if `maybeOf`).

**Comment by goderbauer** [comment]
@gspencergoog is working on removing the nullOk parameters from those other methods you mentioned. We will introduce maybeOf methods for those as well.

**Comment by gspencergoog** [comment]
Yes, see https://flutter.dev/go/eliminating-nullok-parameters

**Comment by Hixie** [comment]
I'm in favour of making every `.of` non-nullable, and adding `.maybeOf` wherever it makes sense. We should create a single migration guide for the whole thing.

**Comment by gspencergoog** [comment]
OK, I'm happy to expand mine to include all the others (I think I have most of them in my doc already).

**Comment by dnfield** [comment]
SGTM. In that case I'm in favor of this approach.

**Comment by goderbauer** [comment]
Looks like we are in agreement on this change. Here's my plan for executing it:

1. I'll submit a PR that adds Directionality.maybeOf to Flutter, leaving Directionality.of unchanged for now.
2. That change rolls into google3.
3. We update flutter_svg to use maybeOf (and bump its minimum flutter version).
4. The updated flutter_svg rolls into google.
5. I update Directionality.of in flutter to be non-null.
6. Profit!

The first step of this plan is https://github.com/flutter/flutter/pull/69117.

@gspencergoog I will add these APIs to https://flutter.dev/go/eliminating-nullok-parameters so we remember to include them in the migration guide.

**Comment by goderbauer** [comment]
This PR no represents step 5 of the plan in my previous post. Submitting this is blocked on submitting https://github.com/dnfield/flutter_svg/pull/435 first.

File: packages/flutter/lib/src/widgets/basic.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/basic.dart:103]: Maybe mention that in debug mode it throws an assertion, and in release mode throws an exception.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

Looks like SVG customer test is still unhappy.

**Comment by goderbauer** [comment]
> Looks like SVG customer test is still unhappy.

Yeah, i still need to submit dnfield/flutter_svg#435 to make those happy.

**Pull request merged**
